src/logger.ts

```diff
@@ -90,48 +84,59 @@ export class SilentLogger implements OptionalLogger {
  *   f(lc2);  // logging inside f will be prefixed with 'foo'
  */
 export class LogContext extends OptionalLoggerImpl {
-  private readonly _s: string;
-  private readonly _logger: OptionalLogger;
+  private readonly _tag: string;
+  private readonly _logSink: LogSink;
+  private readonly _level: LogLevel;
 
-  /**
-   * @param loggerOrLevel If passed a LogLevel a ConsoleLogget is used
-   */
-  constructor(loggerOrLevel: OptionalLogger | LogLevel = 'info', tag = '') {
-    const actualLogger: OptionalLogger = isLogLevel(loggerOrLevel)
-      ? new OptionalLoggerImpl(consoleLogger, loggerOrLevel)
-      : loggerOrLevel;
-
-    super(
-      {
-        log: tag
-          ? (name, ...args) =>
-              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-              actualLogger[name]!(tag, ...args)
-          : (name, ...args) =>
-              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-              actualLogger[name]!(...args),
-      },
-      getLogLevel(actualLogger),
-    );
-    this._s = tag;
-    this._logger = actualLogger;
+  constructor();
+  constructor(level: LogLevel, tag?: string);
+  constructor(logSink: LogSink, level?: LogLevel, tag?: string);
+  constructor(
+    levelOrLogSink?: LogLevel | LogSink,
+    levelOrTag?: LogLevel | string,
+    tag?: string,
+  ) {
+    let logSink: LogSink;
+    let level: LogLevel;
+
+    if (!levelOrLogSink) {
+      // No args: constructor();
+      logSink = consoleLogSink;
+      level = 'info';
+    } else if (isLogLevel(levelOrLogSink)) {
+      // No sink: constructor(level: LogLevel, tag?: string);
+      logSink = consoleLogSink;
+      level = levelOrLogSink;
+      tag = levelOrTag;
+    } else {
+      // constructor(logSink: LogSink, level?: LogLevel, tag?: string);
+      logSink = levelOrLogSink;
+      if (levelOrTag) {
+        level = levelOrTag as LogLevel;
+      } else {
+        level = 'info';
+      }
+      tag ??= '';
+    }
```

Comment:
This is a lot of goop to support all the overrides. Maybe we should get rid of some of them?

Reply:
It is a lot of goop. Can we not require log level and logsink and an optional tag? I find the defaults and overrides to be too much magic, would rather write a tiny wrapper to invoke the constructor with reasonable defaults than to have this stuff in the constructor itself.

Reply:
You could also make them mutable fields so they can be set after construction.

How about:

```ts
constructor(level: string)
tag?: string|undefined;
sinks: LogSink[];
```

Question: why do we have a tag in addition to context? Why not have the tag be the first context?

PS bike shed should be pink.

Reply:
Yeah, tag is just the first context. Can be removed.

Making them public properties is a bit more complicated because then they can change

I'm going to trim down the crazies.